### PR TITLE
Feature/cog 539 implementing additional retriever approaches

### PR DESCRIPTION
**Pull request by hajdul88**
## Summary by CodeRabbit

- **New Features**
	- Introduced a new `CodeSummary` class for improved data handling related to code summaries.
	- Added asynchronous search functionality for retrieving code parts based on description queries.
	- Enhanced `Node` and `Edge` classes with new methods for better data retrieval.

- **Bug Fixes**
	- Improved error handling in the `add_edge` method to log messages instead of raising exceptions.

- **Tests**
	- Removed the `test_add_duplicate_edge` function from the test suite for the `CogneeGraph` class.
	- Added asynchronous unit tests for the `code_description_to_code_part` function, covering various scenarios including input validation and error handling.

### CODE REVIEW

**Comment by hajdul88** [comment]
This is work in progress please do not merge to main!

File: cognee/modules/graph/cognee_graph/CogneeGraphElements.py [hunk 1]
```diff
@@ -109,8 +115,14 @@ def is_edge_alive_in_dimension(self, dimension: int) -> bool:
     def add_attribute(self, key: str, value: Any) -> None:
         self.attributes[key] = value
 
-    def get_attribute(self, key: str, value: Any) -> Union[str, int, float]:
-        return self.attributes[key]
+    def get_attribute(self, key: str) -> Optional[Union[str, int, float]]:
+        return self.attributes.get(key)
+
+    def get_node_from(self):
```
Comment (borisarzentar) [inline cognee/modules/graph/cognee_graph/CogneeGraphElements.py:121]: We usually call these nodes ‘source’ and ‘destination’

File: cognee/modules/retrieval/description_to_codepart_search.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+        for node in results:
+            node_id = str(node.id)
+            node_to_search_from = memory_fragment.get_node(node_id)
+
+            if not node_to_search_from:
+                logging.debug("Node %s not found in memory fragment graph", node_id)
+                continue
+
+            for code_file in node_to_search_from.get_skeleton_neighbours():
```
Comment (borisarzentar) [inline cognee/modules/retrieval/description_to_codepart_search.py:85]: Why we call them ‘skeleton’?

File: cognee/tests/unit/modules/retriever/test_description_to_codepart_search.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        result = await code_description_to_code_part("search query", mock_user, 2)
+
+        assert result == []
+
+
+@pytest.mark.asyncio
+async def test_code_description_to_code_part_invalid_query():
+    """Test that code_description_to_code_part raises ValueError for invalid query."""
+    mock_user = AsyncMock()
+    with pytest.raises(ValueError, match="The query must be a non-empty string."):
```
Comment (borisarzentar) [inline cognee/tests/unit/modules/retriever/test_description_to_codepart_search.py:28]: Give it some space please.

File: cognee/modules/retrieval/description_to_codepart_search.py (hunk 2 above)
Comment (hajdul88) [reply cognee/modules/retrieval/description_to_codepart_search.py:85]: It's from the graph projection implementation from 2 sprints ago. Basically we project only a skeleton of edges, we don't have every properties here, maybe later not even connections.

File: cognee/modules/graph/cognee_graph/CogneeGraphElements.py (hunk 1 above)
Comment (hajdul88) [reply cognee/modules/graph/cognee_graph/CogneeGraphElements.py:121]: will rename them

Comment (hajdul88) [reply cognee/modules/graph/cognee_graph/CogneeGraphElements.py:121]: Done

File: cognee/tests/unit/modules/retriever/test_description_to_codepart_search.py (hunk 3 above)
Comment (hajdul88) [reply cognee/tests/unit/modules/retriever/test_description_to_codepart_search.py:28]: Done

**Review by borisarzentar — APPROVED** [review]

**Pull request merged**